pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/StrictRealtimeSegmentAssignment.java

```diff
@@ -52,86 +58,133 @@
+    List<String> uploadedSegments = new ArrayList<>();
     for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
       // Skip OFFLINE segments as they are not rebalanced, so their assignment in idealState can be stale.
       if (isOfflineSegment(entry.getValue())) {
         continue;
       }
       LLCSegmentName llcSegmentName = LLCSegmentName.of(entry.getKey());
       if (llcSegmentName == null) {
-        nonStandardSegments.add(entry.getKey());
+        uploadedSegments.add(entry.getKey());
         continue;
       }
-      if (llcSegmentName.getPartitionGroupId() == segmentPartitionId) {
-        idealAssignment = entry.getValue().keySet();
-        break;
-      }
-    }
-    if (idealAssignment == null && !nonStandardSegments.isEmpty()) {
-      if (_logger.isDebugEnabled()) {
-        int segmentCnt = nonStandardSegments.size();
-        if (segmentCnt <= 10) {
-          _logger.debug("Check ZK metadata of {} segments: {} for any one also from partition: {}", segmentCnt,
-              nonStandardSegments, segmentPartitionId);
-        } else {
-          _logger.debug("Check ZK metadata of {} segments: {}... for any one also from partition: {}", segmentCnt,
-              nonStandardSegments.subList(0, 10), segmentPartitionId);
-        }
-      }
-      // Check ZK metadata for segments with non-standard LLC segment names to look for a segment that's in the same
-      // table partition with the new segment.
-      for (String nonStandardSegment : nonStandardSegments) {
-        if (SegmentAssignmentUtils.getRealtimeSegmentPartitionId(nonStandardSegment, _tableNameWithType, _helixManager,
-            _partitionColumn) == segmentPartitionId) {
-          idealAssignment = currentAssignment.get(nonStandardSegment).keySet();
-          break;
-        }
+      if (llcSegmentName.getPartitionGroupId() == partitionId) {
+        return entry.getValue().keySet();
       }
     }
-    // Check if the candidate assignment is consistent with idealState. Use idealState if not.
-    if (idealAssignment == null) {
-      _logger.info("No existing assignment from idealState, using the one decided by instancePartitions");
-    } else if (!isSameAssignment(idealAssignment, instancesAssigned)) {
-      _logger.warn("Assignment: {} is inconsistent with idealState: {}, using the one from idealState",
-          instancesAssigned, idealAssignment);
-      instancesAssigned.clear();
-      instancesAssigned.addAll(idealAssignment);
-      if (_controllerMetrics != null) {
-        _controllerMetrics.addMeteredTableValue(_tableNameWithType,
-            ControllerMeter.CONTROLLER_REALTIME_TABLE_SEGMENT_ASSIGNMENT_MISMATCH, 1L);
+    // Check ZK metadata for uploaded segments to look for a segment that's in the same partition
+    for (String uploadedSegment : uploadedSegments) {
+      if (getPartitionId(uploadedSegment) == partitionId) {
```

Comment:
nit: add a comment that there is no need to cache here, as assignSegment() is just one-time shot, but rebalance() can be called in a loop for multiple times